lms/djangoapps/course_api/api.py

```diff
@@ -268,3 +269,65 @@ def get_course_run_url(request, course_id):
     """
     course_run_url = reverse('openedx.course_experience.course_home', args=[course_id])
     return request.build_absolute_uri(course_run_url)
+
+
+def get_course_member_queryset(course_key, include_students=True, access_roles=None, prefetch_accessroles=False, prefetch_enrollments=False):
+    """
+    Returns a User queryset that filters all users related to a course.
+    For example - Students, Teachers, Staffs etc.
+
+    Arguments:
+        course_key (CourseKey): the CourseKey for the course
+        include_students: Wether or not to include students,
+        access_roles:
+            accepts an array of string course access roles.
+            If None provided, it includes all roles.
+        prefetch_accessroles:
+            prefetches CourseAccessRole instances attached with user.
+            This only includes CourseAccessRole instances related to provided CourseKey,
+        prefetch_enrollments:
+            prefetches CourseEnrollment instances attached with user.
+            This only includes CourseEnrollment instances related to provided CourseKey,
+    """
+    queryset = User.objects.filter()
+
+    # if access_roles not given, assign all registered access roles
+    if access_roles is None:
+        access_roles = REGISTERED_ACCESS_ROLES.keys()
+
+    # conditions for filtering based on CourseAccessRole
+    access_role_qs = Q(
+        courseaccessrole__course_id=course_key,
+        courseaccessrole__role__in=access_roles
+    )
+
+    # conditions for filtering based on CourseEnrollment
+    students_qs = Q(
+        courseenrollment__course_id=course_key,
+        courseenrollment__is_active=True
+    )
+
+    if include_students:
+        queryset = queryset.filter(access_role_qs | students_qs)
+    else:
+        queryset = queryset.filter(access_role_qs)
+
+    # prefetch CourseAccessRole items related to the course and given roles
+    if prefetch_accessroles:
+        queryset = queryset.prefetch_related(Prefetch(
+            'courseaccessrole_set',
+            CourseAccessRole.objects.filter(
+                course_id=course_key, role__in=access_roles)
+        ))
+
+    # prefetch CourseEnrollment items related to the course
+    if prefetch_enrollments:
+        queryset = queryset.prefetch_related(Prefetch(
+            'courseenrollment_set',
+            CourseEnrollment.objects.filter(course_id=course_key)
+        ))
+
+    # prevent duplicates
+    queryset = queryset.distinct()
```

Comment:
Run this before doing the prefetch? 

I'm not sure it'll have any impact since django uses lazy querysets, but the code will look better.

Reply:
When querying with in relation to many to many tables, some matches multiple times (ex: both instructor and enrolled) and I've seen it returns duplicates. That's why this has been placed.

Reply:
@shimulch Yes, I got that. I just want you to run `distinct` before the `prefetch` operations. Just move this line to above the prefetch if statement.